code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/MessagingPushUtils.java

```diff
@@ -120,4 +131,126 @@ static int getSmallIconWithResourceName(final String iconName, final Context con
+                    "Failed to get application context. Can't create Uri for cached rich media"
+                            + " file.");
+            return null;
+        }
+
+        final File cachedFile = new File(pathToFile);
+        final String authority = context.getPackageName() + ".fileprovider";
+        return FileProvider.getUriForFile(context, authority, cachedFile);
+    }
+
+    /**
+     * Determines if the push payload contains GIF content in the adb_image field
+     *
+     * @param url the adb_image url
+     * @return true if the adb_image contains a GIF file, false otherwise
+     */
+    static boolean isGifContent(@NonNull final String url) {
+        return url.endsWith(".gif");
+    }
+
+    /**
+     * Retrieves the cached asset from the {@link CacheService} and returns it via a {@code
+     * CompletableFuture}. The asset is retrieved from the cache using the provided key which is the
+     * URL of the asset. The method will wait for the asset to be cached for the specified timeout
+     * duration.
+     *
+     * @param singleThreadScheduledExecutor the {@link Executor} to use for the cached asset
+     *     retrieval
+     * @param key the {@code String} key to retrieve the asset from the cache
+     * @param timeoutInMillis the @code int} timeout in milliseconds to wait for the asset to be
+     *     cached
+     * @return a {@link CompletableFuture} that will be completed with the {@link CacheResult} if
+     *     the asset is cached, or null if the asset is not cached within the timeout
+     */
+    @RequiresApi(api = Build.VERSION_CODES.N)
+    static CompletableFuture<CacheResult> getCachedAsset(
+            @NonNull final Executor singleThreadScheduledExecutor,
+            @NonNull final String key,
+            final int timeoutInMillis) {
+        return CompletableFuture.supplyAsync(
+                () -> tryGetCachedAsset(key, timeoutInMillis), singleThreadScheduledExecutor);
+    }
+
+    /**
+     * Private helper method to retrieve the cached asset from the cache service.
+     *
+     * @param key the {@code String} key to retrieve the asset from the cache
+     * @param timeoutInMillis the {@code int} timeout in milliseconds to wait for the asset to be
+     *     cached
+     * @return the {@link CacheResult} if the asset is cached, or null if the asset is not cached
+     *     within the timeout
+     */
+    @VisibleForTesting
+    static CacheResult tryGetCachedAsset(final @NonNull String key, final int timeoutInMillis) {
+        final String assetCacheLocation = InternalMessagingUtils.getAssetCacheLocation();
+        final CacheService cacheService = ServiceProvider.getInstance().getCacheService();
+        CacheResult cachedAsset = cacheService.get(assetCacheLocation, key);
+        int elapsedTime = 0;
+
+        while (cachedAsset == null && elapsedTime <= timeoutInMillis) {
+            try {
+                Thread.sleep(MessagingConstants.HALF_SECOND_IN_MILLIS);
```

Comment:
Would it be posisble to use CountdownLatch here instead of Thread.sleep?

Reply:
taking a look at this further, there is no callback when the cached asset does become available. using thread.sleep might be necessary as the latch can't be counted down unless the cached asset is checked at a regular interval.